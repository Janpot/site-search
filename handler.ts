export { default } from './src/handler';
export * from './src/handler';
