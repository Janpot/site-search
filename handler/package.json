{
  "source": "../src/handler.ts",
  "main": "../dist/handler.js",
  "module": "../dist/handler.module.js"
}
