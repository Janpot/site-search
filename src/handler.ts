import lunr from 'lunr';
import { IncomingMessage, ServerResponse } from 'http';
import { IndexedDocument } from './types';

let lunrIndex: lunr.Index | undefined;
function getIndex(serializedLunrIndex: object): lunr.Index {
  if (!lunrIndex) {
    lunrIndex = lunr.Index.load(serializedLunrIndex);
  }
  return lunrIndex;
}

type Position = [number, number];

type LevelPosition = {
  position: Position[];
};

type DocMatchPositions = Partial<Record<`l_${number}` | 'text', LevelPosition>>;

type DocMatchesPositions = {
  [match: string]: DocMatchPositions;
};

function mergeTextPositions(
  docMatchPositions: DocMatchesPositions
): Position[] {
  const result: Position[] = [];
  for (const matchPosition of Object.values(docMatchPositions)) {
    const levelMatchPosition = matchPosition.text;
    if (!levelMatchPosition) {
      continue;
    }
    result.push(...levelMatchPosition.position);
  }

  result.sort((a, b) => {
    return a[0] - b[0] || b[1] - a[1];
  });

  return result;
}

interface Snippet {
  parts: string[];
}

export interface BuildSnippetOptions {
  highlights?: Position[];
  // Approximate size of the snippet
  size?: number;
  /// Amount the snippet can grow or shrink on either end to start/stop on a whitespace
  margin?: number;
}

function applyMargin(
  regex: RegExp,
  text: string,
  position: number,
  margin: number
): number {
  let i = 0;
  while (i <= margin) {
    const pre = text.slice(position - i - 1, position - i + 1);
    if (regex.test(pre)) {
      return position - i;
    }
    const post = text.slice(position + i - 1, position + i + 1);
    if (regex.test(post)) {
      return position + i;
    }
    i++;
  }
  return position;
}

export function buildSnippet(
  text: string,
  { highlights = [], size = 100, margin = 10 }: BuildSnippetOptions = {}
): Snippet {
  const pivot = highlights.length > 0 ? highlights[0][0] : 0;
  let sliceStart = pivot - size / 2;
  let sliceEnd = pivot + size / 2;

  if (sliceStart < 0) {
    sliceEnd += -sliceStart;
  } else if (sliceEnd > text.length) {
    sliceStart -= sliceEnd - text.length;
  }

  sliceStart = applyMargin(/.\b\S/, text, sliceStart, margin);
  sliceEnd = applyMargin(/\S\b./, text, sliceEnd, margin);

  const isTruncatedStart = sliceStart > 0;
  const isTruncatedEnd = sliceEnd < text.length;

  sliceStart = Math.max(0, sliceStart);
  sliceEnd = Math.min(sliceEnd, text.length);

  const applicablePositions = highlights.filter(
    ([start, length]) => start >= sliceStart && start + length <= sliceEnd
  );

  const parts = [];
  let i = sliceStart;
  for (const [start, length] of applicablePositions) {
    parts.push(text.slice(i, start));
    parts.push(text.slice(start, start + length));
    i = start + length;
  }

  parts.push(text.slice(i, sliceEnd));

  if (isTruncatedStart) {
    parts[0] = '…' + parts[0];
  }

  if (isTruncatedEnd) {
    parts[parts.length - 1] = parts[parts.length - 1] + '…';
  }

  return { parts };
}

async function search(
  { corpus, index: serializedLunrIndex }: SerializedIndexData,
  query: string
): Promise<SearchApiResult[]> {
  const index = getIndex(serializedLunrIndex);

  const results = index.search(query);
  return results.slice(0, 10).map((result) => {
    const doc = corpus[Number(result.ref)];
    const highlights = mergeTextPositions(
      result.matchData.metadata as DocMatchesPositions
    );

    const snippet = buildSnippet(doc.text, { highlights });

    return {
      hierarchy: doc.hierarchy,
      score: result.score,
      snippet,
    };
  });
}

export interface SerializedIndexData {
  corpus: IndexedDocument[];
  index: object;
}

export interface Options {
  data: SerializedIndexData;
}

export interface SearchApiResult {
  hierarchy: (string | null)[];
  score: number;
  snippet: Snippet;
}

export interface SearchApiResponse {
  results: SearchApiResult[];
}

export default function (options: Options) {
  return async (req: IncomingMessage, res: ServerResponse) => {
    const { searchParams } = new URL(req.url!, 'http://x');
    const query = searchParams.get('q');
    const results = query ? await search(options.data, query + '*') : [];
    res.setHeader('content-type', 'application/json');
    res.end(JSON.stringify({ results }));
  };
}
