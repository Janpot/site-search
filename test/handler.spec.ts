import handler from '../src/handler';
import express from 'express';
import { Server } from 'http';
import fetch from 'node-fetch';
import { readFile } from 'fs/promises';
import { resolve } from 'path';

let server: Server;
const port = 3001;

beforeAll(async () => {
  const index = JSON.parse(
    await readFile(
      resolve(__dirname, './__fixtures__/reactjs/site-search.json'),
      { encoding: 'utf-8' }
    )
  );
  const app = express();
  app.use('/search', handler({ data: index }));
  return new Promise((resolve, reject) => {
    server = app.listen(port).once('listening', resolve).once('error', reject);
  });
});

afterAll(async () => {
  return server?.close();
});

test('handler', async () => {
  const res = await fetch(`http://localhost:${port}/search?q=react`);
  expect(res).toHaveProperty('status', 200);
  const body = await res.json();
  console.log(JSON.stringify(body, null, 2));
  expect(body).toEqual(
    expect.objectContaining({
      results: expect.arrayContaining([]),
    })
  );
});
