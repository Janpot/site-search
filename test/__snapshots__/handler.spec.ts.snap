// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handler 1`] = `
Object {
  "results": Array [
    Object {
      "hierarchy": Array [
        Object {
          "content": "Documentation",
        },
        Object {
          "content": "Implementation Notes",
        },
        null,
        Object {
          "content": "Jumping into the Code ",
        },
        null,
        null,
      ],
      "score": 4.420648571906198,
      "snippet": Object {
        "parts": Array [
          "",
          "ReactMount",
          " is where the code like mountTree() and unmountTree() from this tutorial lives. It takes…",
        ],
      },
    },
    Object {
      "hierarchy": Array [
        Object {
          "content": "Documentation",
        },
        Object {
          "content": "React Top-Level API",
        },
        Object {
          "content": "Reference ",
        },
        Object {
          "content": "React.PureComponent ",
        },
        null,
        null,
      ],
      "score": 2.9187031112896444,
      "snippet": Object {
        "parts": Array [
          "",
          "React.PureComponent",
          " is similar to ",
          "React.Component.",
          " The difference between them is that ",
          "React.Component",
          "…",
        ],
      },
    },
    Object {
      "hierarchy": Array [
        Object {
          "content": "Documentation",
        },
        Object {
          "content": "React Top-Level API",
        },
        Object {
          "content": "Reference ",
        },
        Object {
          "content": "React.PureComponent ",
        },
        null,
        null,
      ],
      "score": 2.9187031112896444,
      "snippet": Object {
        "parts": Array [
          "",
          "React.PureComponent",
          " is similar to ",
          "React.Component.",
          " The difference between them is that ",
          "React.Component",
          "…",
        ],
      },
    },
    Object {
      "hierarchy": Array [
        Object {
          "content": "Documentation",
        },
        Object {
          "content": "ReactDOM",
        },
        Object {
          "content": "Reference ",
        },
        Object {
          "content": "render() ",
        },
        null,
        null,
      ],
      "score": 2.7589419469547867,
      "snippet": Object {
        "parts": Array [
          "Render a ",
          "React",
          " element into the DOM in the supplied container and return a reference to the component…",
        ],
      },
    },
    Object {
      "hierarchy": Array [
        Object {
          "content": "Documentation",
        },
        Object {
          "content": "React Top-Level API",
        },
        Object {
          "content": "Reference ",
        },
        Object {
          "content": "React.Children ",
        },
        Object {
          "content": "React.Children.only ",
        },
        null,
      ],
      "score": 2.7515934777264364,
      "snippet": Object {
        "parts": Array [
          "Verifies that children has only one child (a ",
          "React",
          " element) and returns it. Otherwise this method…",
        ],
      },
    },
    Object {
      "hierarchy": Array [
        Object {
          "content": "Documentation",
        },
        Object {
          "content": "React Top-Level API",
        },
        Object {
          "content": "Reference ",
        },
        Object {
          "content": "React.Children ",
        },
        Object {
          "content": "React.Children.only ",
        },
        null,
      ],
      "score": 2.7515934777264364,
      "snippet": Object {
        "parts": Array [
          "Verifies that children has only one child (a ",
          "React",
          " element) and returns it. Otherwise this method…",
        ],
      },
    },
    Object {
      "hierarchy": Array [
        Object {
          "content": "Documentation",
        },
        Object {
          "content": "ReactDOMServer",
        },
        Object {
          "content": "Reference ",
        },
        Object {
          "content": "renderToStaticNodeStream() ",
        },
        null,
        null,
      ],
      "score": 2.732677973601611,
      "snippet": Object {
        "parts": Array [
          "…except this doesn’t create extra DOM attributes that ",
          "React",
          " uses internally, such as data-",
          "reactroot.",
          " This…",
        ],
      },
    },
    Object {
      "hierarchy": Array [
        Object {
          "content": "Documentation",
        },
        Object {
          "content": "React Top-Level API",
        },
        Object {
          "content": "Reference ",
        },
        Object {
          "content": "React.Suspense ",
        },
        null,
        null,
      ],
      "score": 2.6835520959824595,
      "snippet": Object {
        "parts": Array [
          "",
          "React.Suspense",
          " lets you specify the loading indicator in case some components in the tree below it…",
        ],
      },
    },
    Object {
      "hierarchy": Array [
        Object {
          "content": "Documentation",
        },
        Object {
          "content": "React Top-Level API",
        },
        Object {
          "content": "Reference ",
        },
        Object {
          "content": "React.Suspense ",
        },
        null,
        null,
      ],
      "score": 2.6835520959824595,
      "snippet": Object {
        "parts": Array [
          "",
          "React.Suspense",
          " lets you specify the loading indicator in case some components in the tree below it…",
        ],
      },
    },
    Object {
      "hierarchy": Array [
        Object {
          "content": "Documentation",
        },
        Object {
          "content": "React Top-Level API",
        },
        Object {
          "content": "Reference ",
        },
        Object {
          "content": "React.lazy ",
        },
        null,
        null,
      ],
      "score": 2.558764201864005,
      "snippet": Object {
        "parts": Array [
          "",
          "React.lazy()",
          " lets you define a component that is loaded dynamically. This helps reduce the bundle size…",
        ],
      },
    },
  ],
}
`;
